### DISCUSSION TITLE
Distribution of larger sample sets

### DISCUSSION THREAD

**Discussion by larspalo** [GrandOrgue/grandorgue#1973] [General]
I'm since late May working on a new sample set of a reasonably small organ (15 stops) with a very dry acoustics, but recorded with three different stereo microphone pairs (XY, Near coincident and AB) that will make up my first "surround" sample set. I've realized that even though the "small" size of the organ the sample set will be quite big.

The nice .orgue package format together with wavpacking the samples will of course reduce the distribution size somewhat, but a coarse current estimate is that as a whole it could still easily land somewhere around 9 GB in a single .orgue file.

Now, I begin to wonder if it would make sense to split the sample set into three different .orgue packages, one for each "perspective" that also can be individually loaded. A fourth package could bring the other three together into the whole multi-channel sample set, if that is desired and all three packages are available.

It would reduce the size of each package and in effect reduce the risk of download errors/problems users might encounter with getting it. What are your thoughts about such an approach of distributing a multi channel sample set for GrandOrgue?

**Comment by rousseldenis** [comment]
That could be a good approach IMHO. Some are distributing their sample sets (HW) that way (near in a bundle and surround in another).

Too few people distribute their sample sets with .orgue format, it should be more highlighted.

**Reply by midimusic** [reply]
The computers I use every day have only 8GB of memory with stereo sound. It would be nice to be able to try out your new organ using these computers even if the full feature set is not available. Thus distributing it in several small packages where each package adds additional features would be an extremely good idea.

I currently do not understand why your "surround sound" needs more than two stereo channels. Perhaps you can clarify exactly what you are doing at some later date.

csw900

**Comment by larspalo** [comment]
> Perhaps you can clarify exactly what you are doing at some later date.

Yes. Thanks for your interest in this topic! First off, there's absolutely nothing wrong with your setup! A normal (high quality) stereo sound setup was even considered by Nick Appleton at one point (the blog post is unfortunately lost by now) to be an "upgrade" over a "consumer grade" 5.1 surround audio system. I won't go into that discussion now, but let us just get that out of the way first...

However, multi channel audio system like earlier 5.1 and later 7.1 is now considered "standard" for most "home cinema" audio systems. This has of course also extended to most "consumer grade" audio interfaces for computers. This can be used by GrandOrgue users, if so desired, to route the pipe samples at any "node level" to different audio outputs by means of creating different audio groups.

For creators of pipe organ sample sets the precise "cinema type" of surround audio has never been of particular interest, but they knew that users of any VPO software could utilize multiple audio channels of their audio interfaces at will to route the audio output (of samples) to various speakers. So, instead, many sample set producers have begun to provide their sample sets in multiple "perspectives". That is, when they recorded the organ they didn't only use two microphones (placed somewhere in the church) to create one stereo image of the recording. They instead used many microphones in various stereo setups to capture the emitted audio!

To experiment with this approach I've recently used a Zoom F6 multi track recorder to sample an organ. In effect I've had 6 microphones placed in three different stereo pair setups that recorded the samples of the organ at once.

If we pause for a second here, we must acknowledge that any microphone at any place will receive impulses from the sound emitted that is a "version" of the truth... Two microphones can "collaborate" to create a stereo representation of any sound source, with characteristics that's dependent on their construction and the method of positioning relative to the sound source and each other. I've begun a new screencast series about sample set creation that in some of the videos partially touches on this subject (https://www.youtube.com/playlist?list=PLNz9AJhjubTeNvf379NlcJc5gWlhr8Hc6).

Anyway, for this latest organ recording I had 6 microphones recording the same samples at the same time. One cardioid XY stereo pair placed within the organ case but actually pointing upwards to the ceiling as that was the best option in that particular case to get a fairly even sound of the whole organ. One cardioid "Near coincident" stereo pair fairly close to the front of the organ. And finally, one stereo pair in AB setup of omnidirectional microphones a little bit more away from the front of the organ.

Each pair of microphones creates a stereo image of the organ at that particular position. No one of them is more "correct" than the other, but they are different. They are in fact different "perspectives" of how that organ sounded at that particular time.

When I thus process the samples recorded in this way, this 15 stop instrument becomes like working on a 45 stop instrument as that's basically the amount of work to do. Not to mention that there are multiple releases and also multiple attacks for each stop...

From a practical point of view, and why I started this discussion, it would make sense to distribute this as three separate sample sets that ultimately also can be combined in a multi channel version. Each package would represent one stereo pair placement of the microphones (XY, Near Coincidental and AB). A fourth would utilize all three of them to create a "surround" sample set where the user can decide where to route each stereo pair (or actually even channel if so desired).

Depending on the setup the user has at hands the routing of the channels could be different, but even in a "single" stereo speaker audio setup the multi channel version could provide benefits as the user can decide how much of each of the different microphone pair he/she wants to have representated in the mix. Thus, it ultimately just provides more options for the user to decide how to render the sample set. Each stereo pair has the "truth" from its own "perspective", and by blending them in the proportions the user wants to, he/she has more influence of the rendered sound...

But in more practical terms, dividing a large sample set in multiple files, that each could be downloaded and used on its own, would improve the ratio of download success compared to a single huge file. The .orgue package format allows for utilizing samples from different packages as "dependencies", which would allow to combine multiple package files to a complete multi-channel sample set.

My main question in this topic was if users of GrandOrgue users would approve/accept such a strategy, and think it would be beneficial. It would provide benefits of higher download success ratios and also still keep the easy "installation" procedure - just drop the packages in the appropriate package folder and it would be automatically registered by GO. Also, using the wavpack samples that GO supports would mean much lower download sizes, even though the RAM requirements for the full multi-channel sample set at full quality would still be at original quality amount... For those with less available RAM, all the usual options to reduce the amount of used RAM of course still exist besides the simple option to just use one of the three "perspectives" provided as a stand alone sample set.

**Comment by kerkovits** [comment]
I would definitely agree with the proposed distribution format. I have very limited free disk space on my system and it does not make sense in my particular setup to use a surround sample set.

Eg., I have a version of the Buckeburg organ converted by OdfEdit. It takes ca. 10% of the data partition, although, I only use one stereo pair (other samples are set to 'do not load' to save RAM). This means that ca. 60% of the sample set merely 'wastes' my very limited disk space. The solution you propose would avoid this problem.

**Comment by rousseldenis** [comment]
@larspalo I'm just wondering what is the difference of time between loading plain sample set files or the same in a .orgue file. And the implications on RAM requirements during the extraction. Do you have numbers on that?

**Reply by larspalo** [reply]
I've done some tests with my Bygdsiljum sample set as .orgue file package, un-zipped package and de-compressed (unpacked wavpack files). The actual numbers are of course specifically from my system and I should also point out that the .orgue package file was already registered in GO which otherwise would have take more time at the first load/start of GO. Do note that in no case any cache of the sample set was available for usage, which would have speeded up the loading quite a bit, and between each loading/opening a reboot was done to have a fresh system. The quality settings was in each case uncompressed 24 bit 48 kHz, which is the original quality.

The disc usage of an .orgue package and an un-zipped .orgue package is virtually identical:
![Skärmbild från 2024-08-12 20-57-21](https://github.com/user-attachments/assets/ae16126e-cff0-4d83-8f81-c0f1185e6052)
(compared to the un-zipped version:)
![Skärmbild från 2024-08-12 20-58-00](https://github.com/user-attachments/assets/71bb243b-b0d0-4ddf-a95e-2fe77dd4a457)
(versus the un-zipped and decompressed or unpacked wavpack:)
![Skärmbild från 2024-08-12 20-59-33](https://github.com/user-attachments/assets/5de0c3d5-99c2-4d3a-9dde-0970d69a967a)

And here is the load time/RAM usage progression first for the already registered .orgue file;
![Skärmbild från 2024-08-12 21-13-11](https://github.com/user-attachments/assets/34319242-35e1-4eaa-9ae1-5948448339ff)
And reported RAM usage from GO:
![Skärmbild från 2024-08-12 21-13-26](https://github.com/user-attachments/assets/786eba53-213f-457a-ad33-b3b02928d5bb)

Then the un-zipped package, that's still wavpacked:
![Skärmbild från 2024-08-12 21-21-33](https://github.com/user-attachments/assets/1a497950-6a93-4b71-b439-27c253097351)
And its reported RAM usage from GO:
![Skärmbild från 2024-08-12 21-21-43](https://github.com/user-attachments/assets/815b016c-e467-4acd-92ac-c24bcd204073)

Finally an un-zipped and decompressed or unpacked wavpack version:
![Skärmbild från 2024-08-12 21-27-58](https://github.com/user-attachments/assets/39dfa397-ace1-4c37-88db-8aa419e03fc6)
And its reported RAM usage from GO:
![Skärmbild från 2024-08-12 21-28-12](https://github.com/user-attachments/assets/eae10796-dc31-47a9-bc67-2c74d4f497fc)

My analysis from this would be that the final RAM usage is the same in every case. The processor usage is highest while loading the .orgue package, but it's also, on my system, the fastest (though it's not its first load as it is with all the other cases). The processor usage is clearly the lowest for the de-compressed or un-compressed wavpack samples, but surprisingly the load time is also clearly longer in this particular scenario. Disc usage of an .orgue package or an un-zipped orgue package is virtually the same whereas the un-packed wavpack sample set clearly takes up more disc space.

The overall things get more complicated if one takes into account the re-usability or even worse any modification of the samples themselves. An .orgue package is easy to use for the user but much more difficult to re-use or modify. Any modification basically requires a de-compression first. Re-usage of the samples in the sample set in the package format is possible, but somewhat more difficult. If no adjustments needs to be made to the samples it's possible to reference the files from another .orgue package as they are. An un-zipped and un-packed sample set is obviously the easiest to tamper with, modify or re-use for whatever purpose but it comes at some price of its own too it would seem, at the very least for certain with higher disc usage.
